Deadlock during shutdown

Sorry to report this against an old version, and please resolve if this has been fixed in newer versions which should then urge us to finally upgrade... We're still using 3.16.100v20201030-1916 and are experiencing an occasional deadlock during container shutdown. Here are the two competing threads:

Thread [main] (Suspended)	
	waiting for: AtomicReference<V>  (id=24)	
	Object.wait(long) line: not available [native method]	
	EquinoxBundle$SystemBundle$EquinoxSystemModule(SystemModule).waitForStop(long) line: 173	
	EquinoxBundle$SystemBundle.waitForStop(long) line: 306	
	Equinox.waitForStop(long) line: 217	
	EclipseStarter.shutdown() line: 457	
	EclipseStarter.run(String[], Runnable) line: 274	
	NativeMethodAccessorImpl.invoke0(Method, Object, Object[]) line: not available [native method]	
	NativeMethodAccessorImpl.invoke(Object, Object[]) line: 62	
	DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: 43	
	Method.invoke(Object, Object...) line: 498	
	Main.invokeFramework(String[], URL[]) line: 653	
	Main.basicRun(String[]) line: 590	
	Main.run(String[]) line: 1461	
	Main.main(String[]) line: 1434	

and

Thread [Framework stop - Equinox Container: 8a2dbb34-2363-442e-84e4-4e7b4ea1ff4f] (Suspended)	
	Unsafe.park(boolean, long) line: not available [native method]	
	Unsafe.park(boolean, long) line: 1079	
	LockSupport.park(Object) line: 175	
	ReentrantLock$NonfairSync(AbstractQueuedSynchronizer).parkAndCheckInterrupt() line: 836	
	ReentrantLock$NonfairSync(AbstractQueuedSynchronizer).acquireQueued(AbstractQueuedSynchronizer$Node, int) line: 870	
	ReentrantLock$NonfairSync(AbstractQueuedSynchronizer).acquire(int) line: 1199	
	ReentrantLock$NonfairSync.lock() line: 245	
	ReentrantLock.lock() line: 321	
	ZipBundleFile(CloseableBundleFile<E>).close() line: 340	
	ClasspathEntry.close() line: 257	
	ClasspathManager.close() line: 152	
	EquinoxClassLoader(ModuleClassLoader).close() line: 457	
	BundleLoader.close() line: 359	
	EquinoxContainerAdaptor.invalidateWiring(ModuleWiring, ModuleLoader) line: 264	
	ModuleWiring.invalidate0(boolean) line: 363	
	ModuleWiring.unload() line: 352	
	ModuleContainer.unloadModules() line: 1253	
	ModuleContainer.close() line: 1197	
	EquinoxBundle$SystemBundle$EquinoxSystemModule(SystemModule).stop(Module$StopOptions...) line: 220	
	EquinoxBundle$SystemBundle$EquinoxSystemModule$1.run() line: 220	
	Thread.run() line: 838	


The former waits on the AtomicReference bound to variable stopEvent. The latter tries to obtain the openLock which is held by the former thread. The latter thread would notify the monitor the former thread is waiting on in the finally block that follows in the stop(...) method, but it doesn't reach there because it is hung up at the lock.

This avoids clean shut-down which in turn causes our tests to hang sporadically.

Trying to build I'm getting

``Caused by: java.lang.UnsupportedClassVersionError: org/eclipse/tycho/p2maven/repository/P2ArtifactRepositoryLayout has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0``

What is the minimum JDK version required to build this? Can I be sure to get a bundle JAR that works with Java 8 (byte code version 52)?

--- a/bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/storage/bundlefile/CloseableBundleFile.java
+++ b/bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/storage/bundlefile/CloseableBundleFile.java
@@ -94,7 +94,11 @@ public abstract class CloseableBundleFile<E> extends BundleFile {
 					generation.getBundleInfo().getStorage().getAdaptor().publishContainerEvent(eventType, r.getRevisions().getModule(), e);
 				}
 			}
-			// TODO not sure if throwing a runtime exception is better
+			if (!(e instanceof IOException)) {
+				// throw any unexpected exceptions
+				EquinoxContainer.sneakyThrow(e);
+			}
+			// TODO For IOException, not sure if throwing a runtime exception is better
 			// throw new RuntimeException("Failed to open bundle file.", e);
 			return false;
 		}

--- a/bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/storage/bundlefile/CloseableBundleFile.java
+++ b/bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/storage/bundlefile/CloseableBundleFile.java
@@ -73,17 +73,20 @@ public abstract class CloseableBundleFile<E> extends BundleFile {
 	 * @return true if the bundle file is open and locked
 	 */
 	protected boolean lockOpen() {
+		openLock.lock();
 		try {
-			open(true);
+			internalOpen();
 			return true;
-		} catch (IOException e) {
+		} catch (Throwable e) {
+			// always unlock on any throwable
+			openLock.unlock();
 			if (generation != null) {
 				ModuleRevision r = generation.getRevision();
 				if (r != null) {
 					ContainerEvent eventType = ContainerEvent.ERROR;
 					// If the revision has been removed from the list of revisions then it has been deleted
 					// because the bundle has been uninstalled or updated
-					if (!r.getRevisions().getModuleRevisions().contains(r)) {
+					if (e instanceof IOException && !r.getRevisions().getModuleRevisions().contains(r)) {
 						// instead of filling the log with errors about missing files from
 						// uninstalled/updated bundles just give it an info level
 						eventType = ContainerEvent.INFO;
@@ -105,46 +108,39 @@ public abstract class CloseableBundleFile<E> extends BundleFile {
 	}
 
 	/**
-	 * Opens this bundle file.
-	 * @param keepLock true if the open lock should be retained
-	 * @throws IOException
+	 * Internal method that does the work to open this bundle file. Must hold the
+	 * openLock while calling this method.
+	 * 
 	 */
-	private void open(boolean keepLock) throws IOException {
-		openLock.lock();
-		try {
-			if (closed) {
-				boolean needBackPressure = mruListAdd();
-				if (needBackPressure) {
-					// release lock before applying back pressure
-					openLock.unlock();
-					try {
-						mruListApplyBackPressure();
-					} finally {
-						// get lock back after back pressure
-						openLock.lock();
-					}
-				}
-				// check close again after getting open lock again
-				if (closed) {
-					// always add again if back pressure was applied in case
-					// the bundle file got removed while releasing the open lock
-					if (needBackPressure) {
-						mruListAdd();
-					}
-					// This can throw an IO exception resulting in closed remaining true on exit
-					doOpen();
-					closed = false;
-					if (debug.DEBUG_BUNDLE_FILE_OPEN) {
-						Debug.println("OPENED bundle file - " + toString()); //$NON-NLS-1$
-					}
-				}
-			} else {
-				mruListUse();
-			}
-		} finally {
-			if (!keepLock || closed) {
+	private void internalOpen() throws IOException {
+		if (closed) {
+			boolean needBackPressure = mruListAdd();
+			if (needBackPressure) {
+				// release lock before applying back pressure
 				openLock.unlock();
+				try {
+					mruListApplyBackPressure();
+				} finally {
+					// get lock back after back pressure
+					openLock.lock();
+				}
 			}
+			// check close again after getting open lock again
+			if (closed) {
+				// always add again if back pressure was applied in case
+				// the bundle file got removed while releasing the open lock
+				if (needBackPressure) {
+					mruListAdd();
+				}
+				// This can throw an IO exception resulting in closed remaining true on exit
+				doOpen();
+				closed = false;
+				if (debug.DEBUG_BUNDLE_FILE_OPEN) {
+					Debug.println("OPENED bundle file - " + toString()); //$NON-NLS-1$
+				}
+			}
+		} else {
+			mruListUse();
 		}
 	}
 
@@ -417,7 +413,12 @@ public abstract class CloseableBundleFile<E> extends BundleFile {
 
 	@Override
 	public void open() throws IOException {
-		open(false);
+		openLock.lock();
+		try {
+			internalOpen();
+		} finally {
+			openLock.unlock();
+		}
 	}
 
 	void incrementReference() {

